pyro/nn/auto_reg_nn.py

```diff
@@ -7,6 +7,75 @@
 from torch.nn import functional as F
 
 
+def sample_mask_indices(input_dim, hidden_dim, simple=False, conditional=False):
+    """
+    Samples the indices assigned to hidden units during the construction of MADE masks
+
+    :param input_dim: the dimensionality of the input variable
+    :type input_dim: int
+    :param hidden_dim: the dimensionality of the hidden layer
+    :type hidden_dim: int
+    :param simple: True to sample indices uniformly, false to space indices evenly and round up or down randomly
+    :type simple: bool
+    :param conditional: True if sampling indices for a conditional-MADE, false otherwise
+    :type conditional: bool
+    """
+    start_integer = 0 if conditional else 1
+
+    if simple:
+        # Simple procedure samples each index i.i.d.
+        return torch.randint(start_integer, input_dim, size=(hidden_dim,))
+    else:
+        # "Non-simple" procedure creates fractional indices evenly then rounds at random
+        mk = torch.linspace(start_integer, input_dim - 1, steps=hidden_dim)
+
+        # Randomly decide whether to round up or down indices
+        ints = mk.floor()
+        ints += torch.bernoulli(mk - ints)
+        return ints
+
+
+def create_mask(input_dim, observed_dim, hidden_dim, num_layers, permutation, output_dim_multiplier):
+    """
+    Creates MADE masks for a conditional distribution
+
+    :param input_dim: the dimensionality of the input variable
+    :type input_dim: int
+    :param observed_dim: the dimensionality of the variable that is conditioned on (for conditional densities)
+    :type observed_dim: int
+    :param hidden_dim: the dimensionality of the hidden layer(s)
+    :type hidden_dim: int
+    :param num_layers: the number of hidden layers for which to create masks
+    :type num_layers: int
+    :param permutation: the order of the input variables
+    :type permutation: torch.Tensor
+    :param output_dim_multiplier: tiles the output (e.g. for when a separate mean and scale parameter are desired)
+    :type output_dim_multiplier: int
+    """
+    # Create mask indices for input, hidden layers, and final layer
+    # We use 0 to refer to the elements of the variable being conditioned on,
+    # and range(1:(D_latent+1)) for the input variable
+    var_index = permutation.clone().type(torch.get_default_dtype())
+    var_index[permutation] = torch.arange(input_dim, dtype=torch.get_default_dtype())
```

Comment:
@neerajprad can you please take a look at and comment on the way typing is handled in this PR?

Reply:
I did the `torch.get_default_dtype()` because I need to convert integers to floats but it will be 32- or 64-bit depending on the system I think. Is this consistent with the rest of Pytorch?